### DISCUSSION TITLE
E4 way for Wizards?

### DISCUSSION THREAD

**Discussion by laeubi** [eclipse-platform/eclipse.platform.ui#529] [Q&A]
For E3 there is `org.eclipse.ui.wizards.IWizardRegistry` but there seems no E4 counterpart?

**Comment by vogella** [comment]
Not available so far

**Comment by opcoach** [comment]
In a ideal world, wizards (and also preference pages) should be defined in the application model and contributed in a pure E4 application, using a model fragment...   

An E4 implementation of the IWizardRegistry could be, in this case,  to parse the application model so as to propose the 'new wizard dialog' with all available wizards inside...

**Comment by laeubi** [comment]
Thats excactly what I want, declare a Wizzard in the model (it can then be usead anywhere) with proper injection, and such wizard most probably should then even be defined like a Menu so I have pages I can add, enable/disable and either "direct" or dynamically contributed pages, possibly even things like Windowtitle and so on defined there.

So if there is enough support for such API enhancements I could probably start with an implementation ...

**Comment by opcoach** [comment]
That would be a dream ! :). 

There is an hard work on the model definition... and if it is done, the same must be done for preferences pages... then, we could have fragment to contribute on these categories..

A long time ago (in 2013), you probably remember, there was a 'Dialog' defined in the model. It was supposed to include the different dialogs (Wizard, Property, preference, or any other dialog)... but it was not handy and there was no example to manage it... So it has been removed...

For me, the 'dialog' concept is not really necessary... only preferences, wizard could be added ... may be property pages later... 

For the preferences I already implemented a workaround to make preferences pages in a PURE E4 application.. it is defined in this repo : https://github.com/opcoach/e4Preferences and used by some of my customers.  But this is not included in the model for the moment.. it is miscible with E3 pages definitions.. 

For wizards we can defined them in a PURE E4 application, as this is only JFace level...  but we loose the capability to have the 'New Wizard Dialog' as they are not registered.. in this case they must be opened directly in any handler... 

We could say that evolutions on the application model could drive to the 'Eclipse 5' release... that would be really nice :)

**Reply by laeubi** [reply]
While there is a WizardDialog, Wizards are not really dialogs, so I think it should not be generalized too much and if we really focus on Wizard/Wizardpage I think this can integrate also with the "new/project/import wizards" stuff as the "TYPE" can be a simple attribute, and I think in an RCP app there will be probably other concepts.

I need to investigate a bit on this but will try to start the next days with a first POC...

**Reply by opcoach** [reply]
Yes, putting 'Dialogs' in the model was not a good idea. It must stop on wizards and pages. I totally agree.
